server/core_notification.go

```diff
@@ -0,0 +1,215 @@
+			if err := gob.NewDecoder(bytes.NewReader(cb)).Decode(nc); err != nil {
+				logger.Warn("Could not decode notification cursor.", zap.String("cursor", cursor))
+				return nil, errors.New("Malformed cursor was used.")
+			}
+		}
+	}
+
+	params := []interface{}{userID, limit}
+	cursorQuery := ""
+	if nc.NotificationID != "" {
+		cursorQuery = " AND id > $3::UUID "
+		params = append(params, nc.NotificationID)
+	}
+
+	rows, err := db.Query(`
+SELECT id, subject, content, code, sender_id, create_time
+FROM notification
+WHERE user_id = $1`+cursorQuery+`
+LIMIT $2
+`, params...)
+
+	if err != nil {
+		logger.Error("Could not retrieve notifications.", zap.Error(err))
+		return nil, err
+	}
+
+	notifications := make([]*api.Notification, 0)
+	for rows.Next() {
+		no := &api.Notification{Persistent: true, CreateTime: &timestamp.Timestamp{}}
+		if err := rows.Scan(&no.Id, &no.Subject, &no.Content, &no.Code, &no.SenderId, &no.CreateTime.Seconds); err != nil {
+			logger.Error("Could not scan notification from database.", zap.Error(err))
+			return nil, err
+		}
+		notifications = append(notifications, no)
+	}
+
+	notificationList := &api.NotificationList{}
+	cursorBuf := new(bytes.Buffer)
+	if len(notifications) == 0 {
+		if len(cursor) > 0 {
+			notificationList.CacheableCursor = cursor
+		} else {
+			newCursor := &notificationCacheableCursor{NotificationID: ""}
+			if err := gob.NewEncoder(cursorBuf).Encode(newCursor); err != nil {
+				logger.Error("Could not create new cursor.", zap.Error(err))
+				return nil, err
+			}
+			notificationList.CacheableCursor = base64.RawURLEncoding.EncodeToString(cursorBuf.Bytes())
+		}
+	} else {
+		lastNotification := notifications[len(notifications)-1]
+		newCursor := &notificationCacheableCursor{
+			NotificationID: lastNotification.Id,
+		}
+		if err := gob.NewEncoder(cursorBuf).Encode(newCursor); err != nil {
+			logger.Error("Could not create new cursor.", zap.Error(err))
+			return nil, err
+		}
+		notificationList.Notifications = notifications
+		notificationList.CacheableCursor = base64.RawURLEncoding.EncodeToString(cursorBuf.Bytes())
+	}
+
+	return notificationList, nil
+}
+
+func NotificationDelete(logger *zap.Logger, db *sql.DB, userID uuid.UUID, notificationIDs []string) error {
+	statements := make([]string, 0, len(notificationIDs))
+	params := make([]interface{}, 0, len(notificationIDs))
+	params = append(params, userID)
+
+	for _, id := range notificationIDs {
+		statement := "$" + strconv.Itoa(len(params)+1)
+		statements = append(statements, statement)
+		params = append(params, id)
+	}
+
+	_, err := db.Exec("DELETE FROM notification WHERE user_id = $1 AND id IN ("+strings.Join(statements, ", ")+")", params...)
+	if err != nil {
+		logger.Error("Could not delete notifications.", zap.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+func NotificationSave(logger *zap.Logger, db *sql.DB, notifications map[uuid.UUID][]*api.Notification) error {
+	statements := make([]string, 0, len(notifications))
```

Comment:
Not sure this preallocation is accurate, but no way to handle it without measuring each slice in the map I think.

Reply:
Yep agreed, best effort preallocation at this point I think.